File: skills/cumcm-code-writer/references/algorithm_library/08_统计检验/chi_square_test/matlab/main.m
% 卡方检验（MATLAB 最小可运行模板）
% 独立性检验统计量 chi2 = Σ (O - E)^2 / E，自由度 (r-1)(c-1)；p 值由卡方分布给出。
here = fileparts(mfilename('fullpath'));
addpath(fullfile(here, '..', '..', '..', '_common', 'matlab'));

observed = [42 30 18; 28 45 37];
rowSum = sum(observed, 2); colSum = sum(observed, 1); total = sum(observed(:));
expected = rowSum * colSum / total;
chi2 = sum(sum((observed - expected).^2 ./ expected));
dof = (size(observed, 1) - 1) * (size(observed, 2) - 1);
p = 1 - chi2cdf(chi2, dof);
ratio = mean(expected(:) < 5);

mmlib_report('卡方检验', sprintf('卡方值=%.4f', chi2), sprintf('p值=%.4f', p), ...
    sprintf('自由度=%d', dof), sprintf('期望频数不足5的比例=%.0f%%', ratio * 100), ...
    sprintf('结论=%s', string(p < 0.05)));
T = table(expected(:), 'VariableNames', {'期望频数'});
mmlib_savetable(here, 'chi_square_test', T);

File: skills/cumcm-code-writer/references/algorithm_library/08_统计检验/chi_square_test/python/main.py
"""卡方检验（Python 最小可运行模板）

内容：列联表独立性检验 + 拟合优度检验；同时给出期望频数是否满足 ≥5 的适用性判断。
接口：solve(table) -> (卡方值, p 值, 自由度, 期望频数)；运行：python main.py
"""
from __future__ import annotations

import sys
from pathlib import Path

import numpy as np
from scipy import stats

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "_common" / "python"))
from mmlib import demo, report, save_table  # noqa: E402


def build_demo() -> dict:
    observed = np.array([[42, 30, 18], [28, 45, 37]])
    return {"table": observed, "rows": ["低风险", "高风险"], "cols": ["方案A", "方案B", "方案C"]}


def solve(table):
    table = np.asarray(table, float)
    chi2, p, dof, expected = stats.chi2_contingency(table)
    ratio = float((expected < 5).mean())
    return float(chi2), float(p), int(dof), expected, ratio


if __name__ == "__main__":
    data = demo(__file__, build_demo)
    chi2, p, dof, expected, ratio = solve(data["table"])
    report("卡方检验", 卡方值=round(chi2, 4), p值=f"{p:.4f}", 自由度=dof,
           结论="独立（无关）" if p >= 0.05 else "不独立（相关）",
           期望频数不足5的比例=f"{ratio:.0%}")
    save_table(__file__, "期望频数", {"行": np.repeat(data["rows"], len(data["cols"])),
                                      "列": np.tile(data["cols"], len(data["rows"])),
                                      "期望频数": np.round(expected.ravel(), 4)})
